### ISSUE TITLE
Stale reads?

### ISSUE THREAD

**Issue by stevana** [adjoint-io/raft#41]
Hi,

I've been playing around with the example from your README and I found something that surprises me.

Basically, when I do the following steps:

1. Start node 1
2. Start node 2
3. Start node 3
4. Start client
5. Add all nodes in client
6. set x 8 in client
7. incr x in client
8. Kill and restart one of the nodes
9. read in client

Then sometimes I get back that the value of `x` is 8, which seems wrong? 

I've not been able to reproduce it manually using terminals as described in the README, but rather using some tests that I [wrote](https://github.com/adjoint-io/raft/compare/master...stevana:test/quickcheck-state-machine) (running `stack test --test-arguments="-p test1"` a couple of times should reproduce it).

**Comment by tdietert** [comment]
Hi @stevana, thanks for reporting the issue.

I have read your code, got the gist of what bug it's trying to eek out, and _think_ I have understood why you are seeing these results, why were are only seeing them 1 out of every ~5 runs of the state machine test, and maybe what we can do about it.

The "stale read" issue you are experiencing is related to the paragraph in section 5.3 of the [raft paper](https://raft.github.io/raft.pdf):
> During normal operation, the logs of the leader and
followers stay consistent, so the AppendEntries consistency
check never fails. However, leader crashes can leave
the logs inconsistent (the old leader may not have fully
replicated all of the entries in its log). These inconsistencies
can compound over a series of leader and follower
crashes. Figure 7 illustrates the ways in which followers’
logs may differ from that of a new leader. A follower may
be missing entries that are present on the leader, it may
have extra entries that are not present on the leader, or
both. Missing and extraneous entries in a log may span
multiple terms

What is happening in your test is this:

In `test1`, you are issuing the commands: 
```
    cmds = Commands
      [ Command (SpawnNode 3002) (Set.fromList [ Var 0 ])
      , Command (SpawnNode 3000) (Set.fromList [ Var 1 ])
      , Command (SpawnNode 3001) (Set.fromList [ Var 2 ])
      , Command (Set 8) Set.empty
      , Command Incr Set.empty
      , Command
          (KillNode ( 3000 , Reference (Symbolic (Var 1)) )) Set.empty
      , Command (SpawnNode 3000) (Set.fromList [ Var 3 ])
      , Command Read Set.empty
      ]
```
You expect the last read to return `x := 9` as the state of the system, but instead, quite rarely, you get `x := 8` as the state from the current leader. The reason that you instead get `x := 8` as a response is because of several more granular timing issues of the system:

When you kill node on port 3000, it is sometimes in leader state. Immediately precending your `KillNode` command, you issue an `increment` command to the system and wait for a response. The leader replicates the command on all (or a majority of servers) via an initial append entries RPC, receives a success RPC from each of the followers, responds to the client, *and immediately dies* due to the subsequent `KillNode` command. Immediately following, the node is spawned again (at which point a new leader has been elected), and begins receiving heartbeats from the new leader; at this point, the new leader was _not_ elected having received the second `AppendEntriesRPC` message from the previous leader noting that it should increment it's commit index, thus, it has not updated it's state machine with the committed entries yet.

There are several questions that arise from this, and I will do some more testing in the morning to see if I can narrow down the problem and come up with an answer as to what the work arounds might be, if we need a work around, if this is a bug, and how we should go about fixing it. 

Thanks for the report!

**Comment by tdietert** [comment]
I did, however, discover a bug in the way `AppendEntries` RPCs are constructed when a follower gets behind, so thanks for that! The way I discovered it though was not from an incorrect state machine result not matching the model, but instead the fact that a follower would crash upon receipt of an `AppendEntries` RPC constructed by a leader as the leader performed the log-matching step (see #43)!

**Comment by stevana** [comment]
Thanks for your explanation, I'm not sure if I follow though. Shouldn't the leader hold off responding to the client's write request until all followers have committed and applied the write to their state?

Also, I've added another [example](https://github.com/stevana/raft/commit/10c02f6b181e91cb55c1a7c1ab32d78d5e617198) that causes stale reads. This one doesn't involve killing nodes, merely briefly disabling the posix/io/net/send syscall.

**Comment by tdietert** [comment]
Tomorrow morning I will spend some time investigating the desired system behavior in a scenario like this.

**Comment by stevana** [comment]
>  What is your interpretation of the paragraph from section 5.3 that I pasted in my previous response?

My understanding is that this is an orthogonal problem. Log inconsistencies are resolved during leader election.

>> Shouldn't the leader hold off responding to the client's write request until all followers have committed and applied the write to their state?

> The bold bit is what I'm not sure about.

Have a look at the following paragraph from the paper, the bold bit in particular:

> Once a leader has been elected, it begins servicing
client requests. Each client request contains a command to
be executed by the replicated state machines. The leader
appends the command to its log as a new entry, then is-
sues AppendEntries RPCs in parallel to each of the other
servers to replicate the entry. **When the entry has been
safely replicated (as described below), the leader applies
the entry to its state machine and returns the result of that
execution to the client**.

**Comment by tdietert** [comment]
First, I think we need to clarify what we mean when we talk about _replicating_ and _committing_ a log entry. The two words in do not mean the same thing in the context of the raft paper:

- **replicated**: The log entry has been sent to a majority of followers via an `AppendEntriesRPC` and a majority of followers have responded with "success". _This has to do with the system as a whole_.

- **committed**: The log entry has been applied to the state machine: the `lastAppliedIndex` and `commitIndex` are equal. _This has to do with a single node, not the state of the system as a whole._

I think the problem that we are talking about arises from this distinction. In the current implementation, the leader _does_ respond to the client after a log has been **replicated** on a majority of followers. **However, log replication _does not_ mean that a log has been committed on every follower*, and this is the crux of whether or not what you are testing for is a bug or not.

The way the protocol works in this scenario, with no leader crashes is:

**Leader**:
L1) The leader receives a valid client request to update the replicated state machine
L2) The leader sends the entry to all followers via `AppendEntriesRPC` 
L3) The leader receives successful `AppendEntriesResponseRPC`s from a majority of followers
L4) The leader updates his commit index to the index of the entry 
L5) The leader responds to the client that the entry has been replicated
L6) The leader sends a heartbeat `AppendEntriesRPC` to all followers. 

**Follower**:
F1) The follower receives an `AppendEntriesRPC` from the leader with a new, valid client request
F2) The follower responds with an `AppendEntriesResponseRPC` success to the leader, _assuming the log entry is the valid next entry for the follower as in your tests._ **Notably, the follower does not increment his commit index here**.
F3) The follower receives an `AppendEntriesRPC` from the leader, and increments his commit index and applies the log to his state machine (see final bullet in `AppendEntries RPC` section of protocol description on pg. 4 of the paper).

In the case of your tests, you are killing the leader before `L6` (and thus `F3`) happens, at least sometimes (which is why the test doesn't **always** fail). This means that the log has indeed been replicated (`L1` - `L4`) and that the leader should reply to the client (`L5`). However, since you kill the leader immediately after it responds to the client, sometimes that heartbeat has not been issued and thus the clients do not increment their commit index. This means that when a new leader is elected (one who was previously a follower), the leader has the log entry that was previously replicated, **but his commit index is still not incremented.** 

With my best judgement, I think that the replicated-but-not-committed entry will indeed be committed by all nodes in the network if you allow two heartbeats from the newly elected leader to occur: the log entry has been replicated on all nodes, but the new leader will need to update his `nextIndex` and `matchIndex` for each follower via the first heartbeat message which will result in the leader incrementing his commit index. It is after this _first_ heartbeat that the new leader will accurately report the client request update issued before the previous leader crashed. Next, the leader will broadcast the second heartbeat RPC with its new commit index and all followers will update their commit index accordingly; however, if you kill the leader again, before this second heartbeat message, the same problem will occur as above, and the initial replicated client request will not be committed on all logs until the system is stable with no crashes for at least `2 * (<heartbeat-timeout> + (2 * <network delay>))` milliseconds (which, in practice, is around 100ms).

**With all that said**, this is why I initially thought this scenario was adequately described by section 5.3, notably the paragraph I initially pasted; because it seemingly has so much to do with the timing of the crashes rather than a bug in the implementation. This is also why the observance of this error in the state machine tests occurs regardless of if the leader is crashed or if the network connection is severed. In both cases a new leader is elected before the second heartbeat message that causes all followers to increment their commit indices occurs.

In conclusion, I am not convinced this is a bug, and think that in realistic network environments this won't happen very often at all, and will persist for only a few hundred milliseconds. I think a way to test if my interpretation is correct is to have a timeout of `3 * <heartbeat-timeout>` after killing the leader or severing the network connection and trying to read the system state. This should allow ample time for the commit indices on all nodes to converge; you should never get a stale read no matter how many times you run the tests in this case.

**Comment by stevana** [comment]
>  I am not convinced this is a bug, and think that in realistic network environments this won't happen very often at all, and will persist for only a few hundred milliseconds.

In section 8 of the paper it says:

> Our goal for Raft is to implement linearizable seman-
tics (each operation appears to execute instantaneously,
exactly once, at some point between its invocation and
its response).

From what I understand, if you have linearisable semantics you should never see stale reads (irregardless of timing). The behaviour you're describing sounds more like eventual consistency, which is a weaker guarantee than linearisability.

**Comment by tdietert** [comment]
From the last paragraph of section 8:
> Read-only operations can be handled without writing
anything into the log. However, with no additional measures,
this would run the risk of returning stale data, since
the leader responding to the request might have been superseded
by a newer leader of which it is unaware. Linearizable
reads must not return stale data, and Raft needs
two extra precautions to guarantee this without using the
log. First, a leader must have the latest information on
which entries are committed. **The Leader Completeness
Property guarantees that a leader has all committed entries,
but at the start of its term, it may not know which
those are. To find out, it needs to commit an entry from
its term. _Raft handles this by having each leader commit
a blank no-op entry into the log at the start of its
term._ Second, a leader must check whether it has been deposed
before processing a read-only request (its information
may be stale if a more recent leader has been elected).
Raft handles this by having the leader exchange heartbeat
messages with a majority of the cluster before responding
to read-only requests. Alternatively, the leader
could rely on the heartbeat mechanism to provide a form
of lease [9], but this would rely on timing for safety (it
assumes bounded clock skew).**

To the best of my understanding, the bold sentences describe exactly this problem we are discussing; notably, the last sentence describes the state of the current implmentation, as well as what I wrote in my previous response.

If you agree, then this issue seems to be more of a feature request: We could implement the blocking read request w/ heartbeat method described by the bold & italics sentence. This way, before the new leader responds to your final read request, it would first issue a _second_ heartbeat (it issued the first on becoming leader) and receive a majority success response and thus applying the final `incr x` command to it's state machine, responding with `x := 9` to the client read request.

Currently, the implementation relies on timing for safety, but I do agree that this implementation choice does not perfectly linearisable semantics. **Is this a sensible explanation?** I will heavily consider implementing the blocking read request w/ heartbeat method in the next few days.

**Comment by stevana** [comment]
I'm still not convinced that the paragraph you quote is the problem we're seeing, as they talk about two leaders:

> the leader responding to the request might have been superseded
by a newer leader of which it is unaware.

Which sounds like something that will happen in case of network partitions, while we merely have some packet loss.

Also they say that the safety argument based on timing assumes bounded clock skews, but I don't see how clocks are related to your argument? Under exactly what timing circumstances does your library give linearisable reads? I just tried sleeping two seconds before issuing the read command and I still get a stale value back.

Anyhow, I'm looking forward to being proved wrong by your proposed changes! :-)

**Comment by tdietert** [comment]
Yeah, I agree, I'm not convinced either way... I'm going to spend the morning investigating this potential bug yet again :sweat_smile:

**Comment by tdietert** [comment]
I have indeed discovered the source of the bug. Your tests were instrumental in uncovering a bug that I discovered in previously written concurrency tests that I had removed because I thought the bug existed in the test and not the protocol! :sweat_smile:  

Please check out this PR, merge it into your branch, and double check that you are not observing stale reads any longer: https://github.com/adjoint-io/raft/pull/54

**Comment by tdietert** [comment]
For clarification, my previous comment about PR #54 implemented this: 
> First, a leader must have the latest information on which entries are committed. The Leader Completeness Property guarantees that a leader has all committed entries, but at the start of its term, it may not know which those are. To find out, it needs to commit an entry from its term. Raft handles this by having each leader commit a blank no-op entry into the log at the start of its term.

And then on Friday I implemented the second condition for full linearisability:
>Second, a leader must check whether it has been deposed before processing a read-only request (its information may be stale if a more recent leader has been elected). Raft handles this by having the leader exchange heartbeat messages with a majority of the cluster before responding to read-only requests.

These two changes to the protocol should make it such that your quickcheck state machine tests always succeed :tada:. Thanks for catching the bugs and writing such comprehensive tests!

On that note, if you'd like to contribute the quickcheck state machine test to this library that'd be great! However, I'd ask that you clean up your PR a bit, have some better names for the tests, make sure they run well with a general `stack test` invocation, and other small cleanups. The tests add few dependencies, and makes use of `fiu-X` utils; if you can make this test-suite addition build easily, contain the dependencies to the test suite, and get it working on CI then we'd be happy to merge it.

**Comment by stevana** [comment]
Nice, the tests do indeed pass now -- thanks for debugging and fixing the issue!

I'd be happy to contribute the tests to your library, I'll try to find some time to clean things up soon.

**Comment by tdietert** [comment]
Closing as issue is resolved by PRs #54 and #56
